### PR TITLE
MBS-9297: Set track lengths from CD TOC only if different

### PR DESCRIPTION
**Pull request by yvanzo**
[MBS-9297](https://tickets.metabrainz.org/browse/MBS-9297): Set track lengths from CD TOC only if different

Until then, it was possible to set track durations from CD TOC for any attached medium even if it does not change a millisecond of it.

This patch hides unnecessary links to `/set-durations` and disallows `SetTrackLengths` medium edits with no change.

Note about hidden pre-gap tracks: these are ignored at purpose, because CD TOC is not considered reliable regarding their lengths.

### CODE REVIEW

File: lib/MusicBrainz/Server/Controller/CDTOC.pm [hunk 1]
```diff
@@ -57,9 +57,17 @@ sub show : Chained('load') PathPart('')
     my ($self, $c) = @_;
 
     my $cdtoc = $c->stash->{cdtoc};
+    my $medium_cdtocs = $self->_load_releases($c, $cdtoc);
+
+    for my $mc (@$medium_cdtocs) {
+        ${$mc}{perfect_match} = $c->model('Medium')->perfect_match_cdtoc(
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Controller/CDTOC.pm]: Could we add `perfect_match` as a `Bool` attribute to `Entity::MediumCDTOC`?

File: lib/MusicBrainz/Server/Controller/CDTOC.pm [hunk 2]
```diff
@@ -57,9 +57,17 @@ sub show : Chained('load') PathPart('')
...
+            ${$mc}{medium}{id},
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Controller/CDTOC.pm]: It took me a second to figure out this notation. :) It's a Moose object, right? So `$mc->medium->id` should work. Or at least `$mc->{medium}{id}`.

File: t/lib/t/MusicBrainz/Server/Edit/Medium/SetTrackLengths.pm [hunk 3]
```diff
@@ -39,17 +49,32 @@ test 'Setting track lengths on medium with pregap track' => sub {
...
+        medium_id => 5,
         cdtoc_id => 2
     );
     isa_ok($edit => 'MusicBrainz::Server::Edit::Medium::SetTrackLengths');
 
-    my $medium = $c->model('Medium')->get_by_id(4);
+    my $medium = $c->model('Medium')->get_by_id(5);
     $c->model('Track')->load_for_mediums($medium);
+
+    # Hidden pre-gap track length is untouched at purpose, see above.
```
Comment (mwiencek) [inline t/lib/t/MusicBrainz/Server/Edit/Medium/SetTrackLengths.pm]: Small grammar nitpick: "on purpose" is the right idiom.

File: lib/MusicBrainz/Server/Data/Medium.pm [hunk 4]
```diff
@@ -196,6 +196,30 @@ sub find_for_cdstub {
     );
 }
 
+sub perfect_match_cdtoc
+{
+    my ($self, $medium_id, $cdtoc_id) = @_;
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Data/Medium.pm]: What if this method is passed `$medium` and `$cdtoc` instances with the necessary linked data already loaded? I noticed both subroutines where `perfect_match_cdtoc` is called already need to load those instances. Actually, it might make sense as a `perfect_match` method on `Entity::MediumCDTOC`. Then you wouldn't need to set that attribute in `Controller::CDTOC`. You would just need to create a `MediumCDTOC` instance in `Edit::Medium::SetTrackLengths`.

File: lib/MusicBrainz/Server/Controller/CDTOC.pm (hunk 1 above)
Comment (mwiencek) [inline lib/MusicBrainz/Server/Controller/CDTOC.pm]: Or, maybe you'll agree with making `perfect_match` a method on `Entity::MediumCDTOC`.
 (to replace `Data::Medium::perfect_match_cdtoc`) as described in https://github.com/metabrainz/musicbrainz-server/pull/504#discussion_r111528011, in which case this could be removed.

**Comment by mwiencek** [comment]
LGTM other than possibly moving some code around.

File: lib/MusicBrainz/Server/Controller/CDTOC.pm (hunk 2 above)
Comment (yvanzo) [reply lib/MusicBrainz/Server/Controller/CDTOC.pm]: notation changed in a0c35bd4444a00e3788a1ceccf2439250cfc138c

File: t/lib/t/MusicBrainz/Server/Edit/Medium/SetTrackLengths.pm (hunk 3 above)
Comment (yvanzo) [reply t/lib/t/MusicBrainz/Server/Edit/Medium/SetTrackLengths.pm]: fixed in a0c35bd4444a00e3788a1ceccf2439250cfc138c and 33d0154b923d5c00caf524d473431cda5d511365

**Comment by yvanzo** [comment]
@mwiencek: I reverted the unsuccessful code refactoring (still available in https://github.com/yvanzo/musicbrainz-server/tree/mbs-9297-set-track-lengths-refactoring) and made minor changes to follow your other comments.

File: lib/MusicBrainz/Server/Controller/CDTOC.pm [hunk 5]
```diff
@@ -57,9 +57,17 @@ sub show : Chained('load') PathPart('')
     my ($self, $c) = @_;
 
     my $cdtoc = $c->stash->{cdtoc};
+    my $medium_cdtocs = $self->_load_releases($c, $cdtoc);
+
+    for my $mc (@$medium_cdtocs) {
+        $mc->perfect_match = $c->model('Medium')->perfect_match_cdtoc(
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Controller/CDTOC.pm]: Huh, does that work? I'd expect not since `$mc->perfect_match` is a method call...but I know you can set the new value by passing it to `$mc->perfect_match()`.

File: package.json [hunk 6]
```diff
@@ -39,11 +39,13 @@
     "less-plugin-clean-css": "1.5.1",
     "leven": "2.0.0",
     "lodash": "3.10.1",
+    "lodash-codemods": "^1.0.2",
```
Comment (mwiencek) [inline package.json]: There are some unrelated changes in here and in package-lock.json.

**Comment by mwiencek** [comment]
Not sure what the previous test failure was since the Jenkins build was deleted, but in the interest of closing this I rebased it and implemented my suggestions in https://github.com/metabrainz/musicbrainz-server/pull/504/commits/fa78ea527c50c95b2a92498e2cc4b635283d26c2. Tests pass now (selenium failures are unrelated), and it seems to hide the "set durations" links correctly, so I'll give this an approve. If you want to recheck it to make sure things look okay please do, otherwise let's merge this. :)

**Comment by mwiencek** [comment]
@brainzbot test this please

**Comment by mwiencek** [comment]
@brainzbot test this please

**Comment by mwiencek** [comment]
@brainzbot test this please

**Comment by mwiencek** [comment]
@brainzbot test this please

**Comment by yvanzo** [comment]
Thanks!
